staging/src/k8s.io/apiserver/pkg/server/serve.go

```diff
@@ -85,26 +84,16 @@ func (s *GenericAPIServer) serveSecurely(stopCh <-chan struct{}) error {
 
 	glog.Infof("Serving securely on %s", s.SecureServingInfo.BindAddress)
 	var err error
-	s.effectiveSecurePort, err = RunServer(secureServer, s.SecureServingInfo.BindNetwork, s.ShutdownTimeout, stopCh)
+	s.effectiveSecurePort, err = RunServer(secureServer, s.SecureServingInfo.Listener, s.ShutdownTimeout, stopCh)
 	return err
 }
 
 // RunServer listens on the given port, then spawns a go-routine continuously serving
 // until the stopCh is closed. The port is returned. This function does not block.
-func RunServer(server *http.Server, network string, shutDownTimeout time.Duration, stopCh <-chan struct{}) (int, error) {
-	if len(server.Addr) == 0 {
-		return 0, errors.New("address cannot be empty")
-	}
-
-	if len(network) == 0 {
-		network = "tcp"
+func RunServer(server *http.Server, ln net.Listener, shutDownTimeout time.Duration, stopCh <-chan struct{}) (int, error) {
```

Comment:
we can remove this effectivePort logic from this func. Wondering whether the s.effectiveSecurePort is actually used.

Reply:
I searched the project. and find It only used by test. So I want to remove it and re-implement the test cases using it. Not difficult.

Reply:
With the listener you have the port outside of the bootstrapping, right? So no need to get the port from there anymore.